### PR TITLE
Add ops.getslice for complex indexing by int,slice,None,Ellipsis

### PR DESCRIPTION
**Pull request by fritzo**
Addresses #556
Blocking #553 

This adds `ops.getslice` to perform fancy slice operations like `x[..., 0, ::2, None, :]`, in particular supporting nontrivial slicing, and supporting unsqueezing via None. These fancier slicing ops will be needed for symbolic Gaussians #556. Note this now distinguishes between `ops.getitem` which supports "advanced indexing" where funsors can appear in the index, from `ops.getslice` which supports only indexing of ground values. Numpy distinguishes semantics of these two types of indexing, and it is much easier to fully implement ground indexing (which I believe this PR does for `ops.getitem`).

## Tested
- [x] unit tests
- [x] pass existing tests

### CODE REVIEW

File: funsor/terms.py [hunk 1]
```diff
@@ -1754,6 +1758,20 @@ def eager_getitem_lambda(op, lhs, rhs):
...
 
+@eager.register(Unary, ops.GetsliceOp, Lambda)
+def eager_getslice_lambda(op, x):
+    index = normalize_ellipsis(op.defaults["index"], len(x.shape))
+    head, tail = index[0], index[1:]
+    expr = x.expr
+    if head != slice(None):
+        expr = expr(**{x.var.name: head})
+        if x.var.name not in expr.inputs:
+            return expr
```
Comment (eb8680) [inline funsor/terms.py:1769]: What if `tail` is not empty here?

**Review by eb8680 — COMMENTED** [review]
Code looks good, but can you add a couple of tests for `Lambda` with multiple-value indices to exercise `eager_getslice_lambda`?

File: funsor/terms.py (hunk 1 above)
Comment (fritzo) [reply funsor/terms.py]: thanks, fixed

**Comment by fritzo** [comment]
ok, I've added some simple tests for multiple value indices

**Review by eb8680 — APPROVED** [review]

**Pull request merged**